In [1]:
import numpy as np
import pandas as pd

In [2]:
messages = pd.read_excel('../data/messages_updated.xlsx')
print(messages.shape)

(987, 10)


In [3]:
messages['ДАТА_УСТРАНЕНИЯ_НЕИСПРАВНОСТИ'].isna().sum()

130

In [4]:
# Удалим записи об авариях:

len1 = len(messages)
lst = []

for index, row in messages.iterrows():
    start = row['ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ']
    end = row['ДАТА_УСТРАНЕНИЯ_НЕИСПРАВНОСТИ']

    # с нулевой продолжительностью
    if start == end:
        lst.append(index)
        continue

    # с отрицательной продолжительностью, исключая строки с NaT
    if not pd.isnull(end) and start > end:
        lst.append(index)
        continue

    # M1 продолжительностью до 5 минут, исключая строки с NaT
    if not pd.isnull(end) and end - start <= pd.Timedelta(minutes=5):
            lst.append(index)

messages.drop(lst, inplace=True)

len2 = len(messages)
print('Удалили', len1 - len2, 'строк')

Удалили 17 строк


In [5]:
# Расчет минимального времени устранения неисправности

timedelta = messages[['ИМЯ_МАШИНЫ',
                      'НАЗВАНИЕ_ТЕХ_МЕСТА',
                      'ВИД_СООБЩЕНИЯ',
                      'ОПИСАНИЕ', 
                      'ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ',
                      'ДАТА_УСТРАНЕНИЯ_НЕИСПРАВНОСТИ',
                      'ТЕКСТ_ГРУППЫ_КОДОВ']].copy()

timedelta['ПРОДОЛЖИТЕЛЬНОСТЬ'] = \
    timedelta['ДАТА_УСТРАНЕНИЯ_НЕИСПРАВНОСТИ'] - timedelta['ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ']

#timedelta.sort_values(by=['ПРОДОЛЖИТЕЛЬНОСТЬ'], ascending=True).to_excel('../data/timedelta.xlsx')

In [6]:
# Минимальная продолжительность M1

M1_min = timedelta[timedelta['ВИД_СООБЩЕНИЯ'] == 'M1']['ПРОДОЛЖИТЕЛЬНОСТЬ'].min()
timedelta[timedelta['ВИД_СООБЩЕНИЯ'] == 'M1'].sort_values(by=['ПРОДОЛЖИТЕЛЬНОСТЬ'], ascending=True).head()

ИМЯ_МАШИНЫ                        НАЗВАНИЕ_ТЕХ_МЕСТА   
986  ЭКСГАУСТЕР А/М №9        ЭЛЕКТРООБОРУДОВАНИЯ ЭКСГАУСТЕРА №9  \
339  ЭКСГАУСТЕР А/М №7                         ЭКСГАУСТЕР А/М №7   
176  ЭКСГАУСТЕР А/М №4             ПОДШИПНИК ОПОРНЫЙ №2 ЭКСГ. №4   
892  ЭКСГАУСТЕР А/М №5  ТИРИСТ. ВОЗБУДИТЕЛЬ ВТ-РЭМ-400 ЭКСГ5 ВУ1   
692  ЭКСГАУСТЕР А/М №4                ЭЛЕКТРОАППАРАТУРА ЭКСГ. №4   

    ВИД_СООБЩЕНИЯ                                ОПИСАНИЕ   
986            M1            остановка эксг. №9 (2РУ МТЗ)  \
339            M1          остановка эксгаустера №7 (МТЗ)   
176            M1     остановка эксгаустера (температура)   
892            M1            остановка эк-ра №5 (2РУ МТЗ)   
692            M1  остановка экс-ра №4 (РТ газ. задвижки)   

    ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ ДАТА_УСТРАНЕНИЯ_НЕИСПРАВНОСТИ   
986       2021-12-29 16:27:02           2021-12-29 16:33:03  \
339       2020-01-10 19:38:18           2020-01-10 19:44:19   
176       2019-07-30 19:14:34           2019-07-30 19:20:35   
892       2021-09-10 23:18:08           2021-09-10 23:25:09   
692       2021-02-17 19:50:34           2021-02-17 19:57:35   

             ТЕКСТ_ГРУППЫ_КОДОВ ПРОДОЛЖИТЕЛЬНОСТЬ  
986      ТЕХНИЧЕСКИЕ НЕПЛАНОВЫЕ   0 days 00:06:01  
339  ТЕХНОЛОГИЧЕСКИЕ НЕПЛАНОВЫЕ   0 days 00:06:01  
176      ТЕХНИЧЕСКИЕ НЕПЛАНОВЫЕ   0 days 00:06:01  
892      ТЕХНИЧЕСКИЕ НЕПЛАНОВЫЕ   0 days 00:07:01  
692      ТЕХНИЧЕСКИЕ НЕПЛАНОВЫЕ   0 days 00:07:01

In [7]:
# Минимальная продолжительность M3

M3_min = timedelta[timedelta['ВИД_СООБЩЕНИЯ'] == 'M3']['ПРОДОЛЖИТЕЛЬНОСТЬ'].min()
timedelta[timedelta['ВИД_СООБЩЕНИЯ'] == 'M3'].sort_values(by=['ПРОДОЛЖИТЕЛЬНОСТЬ'], ascending=True).head()

ИМЯ_МАШИНЫ          НАЗВАНИЕ_ТЕХ_МЕСТА ВИД_СООБЩЕНИЯ   
935  ЭКСГАУСТЕР А/М №6  ТР-Р ТМ-4000-10/6 ЭКСГ. №6            M3  \
947  ЭКСГАУСТЕР А/М №9  ТР-Р ТМ-6300-10/6 ЭКСГ. №9            M3   
946  ЭКСГАУСТЕР А/М №9  ТР-Р ТМ-6300-10/6 ЭКСГ. №9            M3   
945  ЭКСГАУСТЕР А/М №9  ТР-Р ТМ-6300-10/6 ЭКСГ. №9            M3   
944  ЭКСГАУСТЕР А/М №9  ТР-Р ТМ-6300-10/6 ЭКСГ. №9            M3   

                                ОПИСАНИЕ ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ   
935                      Нагрев контакта       2021-11-08 12:42:00  \
947          Подтеки масла на изоляторах       2021-11-23 15:35:00   
946  Подгар контактов термосигнализатора       2021-11-23 15:33:00   
945         Низкий уровень масла в тр-ре       2021-11-23 15:32:00   
944       Термосигнализатор не закреплен       2021-11-23 15:31:00   

    ДАТА_УСТРАНЕНИЯ_НЕИСПРАВНОСТИ ТЕКСТ_ГРУППЫ_КОДОВ ПРОДОЛЖИТЕЛЬНОСТЬ  
935           2021-11-08 12:50:00                NaN   0 days 00:08:00  
947           2021-11-23 15:45:45                NaN   0 days 00:10:45  
946           2021-11-23 15:45:38                NaN   0 days 00:12:38  
945           2021-11-23 15:45:32                NaN   0 days 00:13:32  
944           2021-11-23 15:45:25                NaN   0 days 00:14:25

In [8]:
# Округление дат

messages['ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ'] = messages['ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ'].apply(lambda t: t.ceil('10S'))

messages['ДАТА_УСТРАНЕНИЯ_НЕИСПРАВНОСТИ'] = messages['ДАТА_УСТРАНЕНИЯ_НЕИСПРАВНОСТИ'].apply(
    lambda t: t.floor('10S') if not pd.isnull(t) else np.nan  # чтобы не округлять NaT
)
messages.head()

Unnamed: 0              МАШИНА         ИМЯ_МАШИНЫ   
0         390         AA2/006-006  ЭКСГАУСТЕР А/М №9  \
1         391  CH-AGP-AG2/011-005  ЭКСГАУСТЕР А/М №9   
2         392         AA2/002-006  ЭКСГАУСТЕР А/М №5   
3         393  CH-AGP-AG2/003-005  ЭКСГАУСТЕР А/М №5   
4         394  CH-AGP-AG2/003-005  ЭКСГАУСТЕР А/М №5   

                    ТЕХ_МЕСТО                   НАЗВАНИЕ_ТЕХ_МЕСТА   
0         AA2/006-006-002-008     ЗАПОРНАЯ АРМАТУРА ЭКСГАУСТЕРА №9  \
1      CH-AGP-AG2/011-005-002  МАСЛОСТАНЦИЯ ЖИДКОЙ СМАЗКИ ЭКСГ. №9   
2         AA2/002-006-002-004     МАСЛООХЛАДИТЕЛЬ  М-05-1 ЭКСГ. №5   
3  CH-AGP-AG2/003-005-002-004     МАСЛООХЛАДИТЕЛЬ  М-05-1 ЭКСГ. №5   
4  CH-AGP-AG2/003-005-002-004     МАСЛООХЛАДИТЕЛЬ  М-05-1 ЭКСГ. №5   

  ВИД_СООБЩЕНИЯ                                  ОПИСАНИЕ   
0            M3                      неисправен двигатель  \
1            M3                      неисправен двигатель   
2            M3  менять прокладку на подачу воды в охлади   
3            M3  менять прокладку на подачу воды в охлади   
4            M3  менять прокладку на подачу воды в охлади   

  ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ ДАТА_УСТРАНЕНИЯ_НЕИСПРАВНОСТИ ТЕКСТ_ГРУППЫ_КОДОВ  
0       2019-01-21 00:00:00           2019-02-25 00:00:00                NaN  
1       2019-01-21 12:26:10           2019-02-25 00:00:00                NaN  
2       2019-01-21 15:35:30           2019-01-23 11:00:00                NaN  
3       2019-01-21 16:33:00           2019-01-22 00:00:00                NaN  
4       2019-01-21 16:36:10           2019-06-04 15:35:20                NaN

**Анализ данных, не содержащих сведений о дате завершения неисправности**  
Видим, что 130 записей о M3 ошибочно не завершаются никогда. Это характерно для всех эксгаустеров, кроме №4, уже в первый же год (2019) наблюдений. Таким образом, большая часть датасета неправильно размечена по M3.

In [9]:
messages['ДАТА_УСТРАНЕНИЯ_НЕИСПРАВНОСТИ'].isna().sum()

130

In [20]:
endless_fail = messages[messages['ДАТА_УСТРАНЕНИЯ_НЕИСПРАВНОСТИ'].isna()].copy()
endless_fail['ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ'].dt.year.value_counts()

ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ
2021    70
2020    37
2019    23
Name: count, dtype: int64

In [21]:
endless_fail.groupby(['ИМЯ_МАШИНЫ', endless_fail['ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ'].dt.year]).size()

ИМЯ_МАШИНЫ         ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ
ЭКСГАУСТЕР А/М №4  2020                          9
                   2021                         17
ЭКСГАУСТЕР А/М №5  2019                          2
                   2020                          6
                   2021                         11
ЭКСГАУСТЕР А/М №6  2019                          3
                   2020                          5
                   2021                         10
ЭКСГАУСТЕР А/М №7  2019                         13
                   2020                          9
                   2021                          7
ЭКСГАУСТЕР А/М №8  2019                          2
                   2020                          7
                   2021                         12
ЭКСГАУСТЕР А/М №9  2019                          3
                   2020                          1
                   2021                         13
dtype: int64

In [22]:
endless_fail[endless_fail['ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ'].dt.year == 2019]

Unnamed: 0              МАШИНА         ИМЯ_МАШИНЫ   
87          501  CH-AGP-AG2/009-005  ЭКСГАУСТЕР А/М №8  \
164         591  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
189         616  CH-AGP-AG2/003-005  ЭКСГАУСТЕР А/М №5   
199         626  CH-AGP-AG2/011-005  ЭКСГАУСТЕР А/М №9   
200         627  CH-AGP-AG2/011-005  ЭКСГАУСТЕР А/М №9   
243         680  CH-AGP-AG2/003-005  ЭКСГАУСТЕР А/М №5   
244         681  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
245         682  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
248         686  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
253         697  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
254         698  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
262         707  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
275         723  CH-AGP-AG2/005-005  ЭКСГАУСТЕР А/М №6   
284         734  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
292         746  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
293         747  CH-AGP-AG2/005-005  ЭКСГАУСТЕР А/М №6   
297         754  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
299         757  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
301         760  CH-AGP-AG2/005-005  ЭКСГАУСТЕР А/М №6   
318         796  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
319         797  CH-AGP-AG2/007-005  ЭКСГАУСТЕР А/М №7   
321         800  CH-AGP-AG2/009-005  ЭКСГАУСТЕР А/М №8   
322         801  CH-AGP-AG2/011-005  ЭКСГАУСТЕР А/М №9   

                          ТЕХ_МЕСТО                       НАЗВАНИЕ_ТЕХ_МЕСТА   
87       CH-AGP-AG2/009-005-002-004         МАСЛООХЛАДИТЕЛЬ  М-05-1 ЭКСГ. №8  \
164      CH-AGP-AG2/007-005-001-002                           РОТОР ЭКСГ. №7   
189      CH-AGP-AG2/003-005-003-002          РЕДУКТОР ГАЗ. ЗАДВИЖКИ ЭКСГ. №5   
199      CH-AGP-AG2/011-005-002-001              МАСЛОНАСОС РАБОЧИЙ ЭКСГ. №9   
200      CH-AGP-AG2/011-005-001-003        ПОДШИПНИК ОПОРНО-УПОРНЫЙ ЭКСГ. №9   
243  CH-AGP-AG2/003-005-001-005-002                     ПОДШИПНИК ОПОРНЫЙ №2   
244      CH-AGP-AG2/007-005-001-003        ПОДШИПНИК ОПОРНО-УПОРНЫЙ ЭКСГ. №7   
245  CH-AGP-AG2/007-005-001-005-002                     ПОДШИПНИК ОПОРНЫЙ №2   
248      CH-AGP-AG2/007-005-001-004               ПОДШИПНИК ОПОРНЫЙ ЭКСГ. №7   
253      CH-AGP-AG2/007-005-001-004               ПОДШИПНИК ОПОРНЫЙ ЭКСГ. №7   
254  CH-AGP-AG2/007-005-001-005-002                     ПОДШИПНИК ОПОРНЫЙ №2   
262      CH-AGP-AG2/007-005-001-004               ПОДШИПНИК ОПОРНЫЙ ЭКСГ. №7   
275      CH-AGP-AG2/005-005-001-001                          КОРПУС ЭКСГ. №6   
284      CH-AGP-AG2/007-005-001-004               ПОДШИПНИК ОПОРНЫЙ ЭКСГ. №7   
292      CH-AGP-AG2/007-005-001-004               ПОДШИПНИК ОПОРНЫЙ ЭКСГ. №7   
293      CH-AGP-AG2/005-005-001-003        ПОДШИПНИК ОПОРНО-УПОРНЫЙ ЭКСГ. №6   
297      CH-AGP-AG2/007-005-001-004               ПОДШИПНИК ОПОРНЫЙ ЭКСГ. №7   
299  CH-AGP-AG2/007-005-001-005-003  ЭЛЕКТРОДВИГАТЕЛЬ ДСПУ-140-84-4 ЭКСГ. №7   
301  CH-AGP-AG2/005-005-001-005-002                     ПОДШИПНИК ОПОРНЫЙ №2   
318      CH-AGP-AG2/007-005-001-004               ПОДШИПНИК ОПОРНЫЙ ЭКСГ. №7   
319  CH-AGP-AG2/007-005-001-005-002                     ПОДШИПНИК ОПОРНЫЙ №2   
321      CH-AGP-AG2/009-005-001-002                           РОТОР ЭКСГ. №8   
322      CH-AGP-AG2/011-005-001-002                           РОТОР ЭКСГ. №9   

    ВИД_СООБЩЕНИЯ                    ОПИСАНИЕ ДАТА_НАЧАЛА_НЕИСПРАВНОСТИ   
87             M3          греются подшипники       2019-04-08 21:50:50  \
164            M3                  течь масла       2019-06-18 18:31:20   
189            M3                  нет смазки       2019-08-13 16:38:00   
199            M3               неисправность       2019-08-21 19:32:00   
200            M3                утечка масла       2019-08-21 19:40:40   
243            M3                утечка масла       2019-09-30 17:17:30   
244            M3                           у       2019-09-30 17:18:50   
245            M3                утечка масла       2019-09-30 17:19:40   
248            M

In [10]:
# Сохранение для дальнейшей работы
messages.to_excel('../data/processed/messages_processed.xlsx', index=False)